src/QsCompiler/Transformations/CallGraph/CallGraphBase.cs

```diff
@@ -0,0 +1,255 @@
+            ImmutableArray<KeyValuePair<TNode, TEdge>>.Empty
+            .ToLookup(kvp => kvp.Key, kvp => kvp.Value);
+
+        // Member Fields
+
+        /// <summary>
+        /// This is a dictionary mapping source nodes to information about target nodes. This information is represented
+        /// by a dictionary mapping target node to the edges pointing from the source node to the target node.
+        /// </summary>
+        protected readonly Dictionary<TNode, Dictionary<TNode, ImmutableArray<TEdge>>> dependencies =
+            new Dictionary<TNode, Dictionary<TNode, ImmutableArray<TEdge>>>();
+
+        // Properties
+
+        /// <summary>
+        /// The number of nodes in the call graph.
+        /// </summary>
+        public int Count => this.dependencies.Count;
+
+        /// <summary>
+        /// A hash set of the nodes in the call graph.
+        /// </summary>
+        public ImmutableHashSet<TNode> Nodes => this.dependencies.Keys.ToImmutableHashSet();
+
+        // Member Methods
+
+        /// <summary>
+        /// Returns true if the given node is found in the call graph, false otherwise.
+        /// Throws ArgumentNullException if argument is null.
+        /// </summary>
+        public bool ContainsNode(TNode node)
+        {
+            if (node is null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            return this.dependencies.ContainsKey(node);
+        }
+
+        /// <summary>
+        /// Returns the children nodes of a given node. Each key in the returned lookup is a child
+        /// node of the given node. Each value in the lookup is an edge connecting the given node to
+        /// the child node represented by the associated key.
+        /// Returns an empty ILookup if the node was found with no dependencies or was not found in
+        /// the graph.
+        /// Throws ArgumentNullException if argument is null.
+        /// </summary>
+        public ILookup<TNode, TEdge> GetDirectDependencies(TNode node)
+        {
+            if (node is null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            if (this.dependencies.TryGetValue(node, out var dep))
+            {
+                return dep
+                    .SelectMany(kvp => kvp.Value, Tuple.Create)
+                    .ToLookup(tup => tup.Item1.Key, tup => tup.Item2);
+            }
+            else
+            {
+                return EmptyDependency();
+            }
+        }
+
+        /// <summary>
+        /// Adds a dependency to the call graph using the two nodes and the edge between them.
+        /// The nodes are added to the graph if they are not already there. The edge is always added.
+        /// Throws ArgumentNullException if any of the arguments are null.
+        /// </summary>
+        protected void AddDependency(TNode fromNode, TNode toNode, TEdge edge)
```

Comment:
Is the call graph intended to be immutable or mutable?

Reply:
I was trying to do the best of both worlds: I need parts of it to be mutable while I'm building it up with the walkers, but after it is built, it should be immutable. That is why the properties were immutable datatypes. Was there something specific to AddDependency regarding this, or is this just a general comment for the call graph?

Reply:
Oh I see how it is related. It adds dependencies to the graph 😁!

Reply:
The access level for this could be set to `private protected` which would mean it was only accessible in this assembly and only to its derived classes. But I'm not sure this is the thing to do, as anyone who wants to create their own call graph class derived from this class would need some way to being able to build that graph, which would be this method.

Reply:
You could use the builder pattern, where all the members that mutate the graph are private, and can only be accessed through a nested builder class that returns an immutable graph after you're done building it.

Reply:
I guess that is kind of what the walkers are (the builder class), except they are not nested, and nesting them could become pretty messy.

Reply:
I just had in mind only exposing `AddDependency` and `AddNode` in the builder, not adding any additional logic. I guess they wouldn't need to be private methods - they could probably directly be public methods in the builder class.

Reply:
The builder classes would have as methods on them the AddDependency and AddNode public methods, and they would have access to the private mutable fields of the call graph by virtue of the builder being a nested class. The walkers would then use the methods of the builder class to build the call graph. Is that correct? I'm not really sure what this does other than put all the mutating methods into a public nested class.

Reply:
Ah so when you instantiate the builder class object, you have access to the mutator methods. But when you instantiate the regular call graph class, you don't have access to those anymore, so it is immutable, right?

Reply:
Yes, the builder would have a method that returns an immutable graph based on what has been added to the builder. The builder couldn't be used to mutate the returned graph. That enforces that the graph can only be mutated while it is being built.

Reply:
Removed the CallGraphBase as an inherited type and changed it to a call graph builder class. The Simple and Concrete call graph types have a private builder member that is used during construction.